{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "from pyteomics import fasta\n",
    "import random\n",
    "\n",
    "import os\n",
    "import sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "module_path = os.path.abspath(os.path.join('../..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "    \n",
    "from src import runner"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hacky False Positive Calculation\n",
    "For the sake of making it easier, we are going to do the FP in a bit of a weird way. We have a dataset with 280 proteins we know to be the acutal source. We will run hypedsearch with both this set and a set of 300 prots that are NOT in this set. Then we will compare results. \n",
    "1. Load in the truth set\n",
    "2. Create another set of proteins that are exclusive of the truth set\n",
    "3. Run hypedsearch on both sets\n",
    "4. Count the number of results that the non truth set had over the truth set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "specPath = '/Users/zacharymcgrath/Desktop/nod2 data/filteredSpec/'\n",
    "\n",
    "truth_set = '/Users/zacharymcgrath/Desktop/nod2 data/filteredNOD2.fasta'\n",
    "not_truth_set = '/Users/zacharymcgrath/Desktop/nod2 data/not_truth_subset.fasta'\n",
    "whole_db = '/Users/zacharymcgrath/Desktop/nod2 data/all data/NOD2_mouse_database.fasta'\n",
    "\n",
    "outputDir = '/Users/zacharymcgrath/Desktop/Experiment output/FP/'\n",
    "minPep = 3\n",
    "maxPep = 30\n",
    "tolerance = 20\n",
    "relative_abundance_filter = 0.0\n",
    "precursor_tolerance = 3\n",
    "peak_filter = 25\n",
    "verbose = True\n",
    "\n",
    "truth_run_params = {\n",
    "    'spectra_folder': specPath,\n",
    "    'database_file': truth_set,\n",
    "    'output_dir': outputDir + 'truth/',\n",
    "    'min_peptide_len': minPep,\n",
    "    'max_peptide_len': maxPep,\n",
    "    'tolerance': tolerance,\n",
    "    'precursor_tolerance': precursor_tolerance,\n",
    "    'peak_filter': peak_filter, \n",
    "    'relative_abundance_filter': relative_abundance_filter,\n",
    "    'verbose': verbose,\n",
    "    'DEBUG': False\n",
    "}\n",
    "\n",
    "non_truth_run_params = {\n",
    "    'spectra_folder': specPath,\n",
    "    'database_file': not_truth_set,\n",
    "    'output_dir': outputDir + 'not_truth/',\n",
    "    'min_peptide_len': minPep,\n",
    "    'max_peptide_len': maxPep,\n",
    "    'tolerance': tolerance,\n",
    "    'precursor_tolerance': precursor_tolerance,\n",
    "    'peak_filter': peak_filter, \n",
    "    'relative_abundance_filter': relative_abundance_filter,\n",
    "    'verbose': verbose, \n",
    "    'DEBUG': False\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Load in truth set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_name = lambda x: x.split('|')[-1].split(' ')[0]\n",
    "\n",
    "ts = {get_name(entry.description): None for entry in fasta.read(truth_set)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Create the other set of 280 proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\r",
      "choosing prot 1/280\r",
      "choosing prot 2/280\r",
      "choosing prot 3/280\r",
      "choosing prot 4/280\r",
      "choosing prot 5/280\r",
      "choosing prot 6/280\r",
      "choosing prot 7/280\r",
      "choosing prot 8/280\r",
      "choosing prot 9/280\r",
      "choosing prot 10/280\r",
      "choosing prot 11/280\r",
      "choosing prot 12/280\r",
      "choosing prot 13/280\r",
      "choosing prot 14/280\r",
      "choosing prot 15/280\r",
      "choosing prot 16/280\r",
      "choosing prot 17/280\r",
      "choosing prot 18/280\r",
      "choosing prot 19/280\r",
      "choosing prot 20/280\r",
      "choosing prot 21/280\r",
      "choosing prot 22/280\r",
      "choosing prot 23/280\r",
      "choosing prot 24/280\r",
      "choosing prot 25/280\r",
      "choosing prot 26/280\r",
      "choosing prot 27/280\r",
      "choosing prot 28/280\r",
      "choosing prot 29/280\r",
      "choosing prot 30/280\r",
      "choosing prot 31/280\r",
      "choosing prot 32/280\r",
      "choosing prot 33/280\r",
      "choosing prot 34/280\r",
      "choosing prot 35/280\r",
      "choosing prot 36/280\r",
      "choosing prot 37/280\r",
      "choosing prot 38/280\r",
      "choosing prot 39/280\r",
      "choosing prot 40/280\r",
      "choosing prot 41/280\r",
      "choosing prot 42/280\r",
      "choosing prot 43/280\r",
      "choosing prot 44/280\r",
      "choosing prot 45/280\r",
      "choosing prot 46/280\r",
      "choosing prot 47/280\r",
      "choosing prot 48/280\r",
      "choosing prot 49/280\r",
      "choosing prot 50/280\r",
      "choosing prot 51/280\r",
      "choosing prot 52/280\r",
      "choosing prot 53/280\r",
      "choosing prot 54/280\r",
      "choosing prot 55/280\r",
      "choosing prot 56/280\r",
      "choosing prot 57/280\r",
      "choosing prot 58/280\r",
      "choosing prot 59/280\r",
      "choosing prot 60/280\r",
      "choosing prot 61/280\r",
      "choosing prot 62/280\r",
      "choosing prot 63/280\r",
      "choosing prot 64/280\r",
      "choosing prot 65/280\r",
      "choosing prot 66/280\r",
      "choosing prot 67/280\r",
      "choosing prot 68/280\r",
      "choosing prot 69/280\r",
      "choosing prot 70/280\r",
      "choosing prot 71/280\r",
      "choosing prot 72/280\r",
      "choosing prot 73/280\r",
      "choosing prot 74/280\r",
      "choosing prot 75/280\r",
      "choosing prot 76/280\r",
      "choosing prot 77/280\r",
      "choosing prot 78/280\r",
      "choosing prot 79/280\r",
      "choosing prot 80/280\r",
      "choosing prot 81/280\r",
      "choosing prot 82/280\r",
      "choosing prot 83/280\r",
      "choosing prot 84/280\r",
      "choosing prot 85/280\r",
      "choosing prot 86/280\r",
      "choosing prot 87/280\r",
      "choosing prot 88/280\r",
      "choosing prot 89/280\r",
      "choosing prot 90/280\r",
      "choosing prot 91/280\r",
      "choosing prot 92/280\r",
      "choosing prot 93/280\r",
      "choosing prot 94/280\r",
      "choosing prot 95/280\r",
      "choosing prot 96/280\r",
      "choosing prot 97/280\r",
      "choosing prot 98/280\r",
      "choosing prot 99/280\r",
      "choosing prot 100/280\r",
      "choosing prot 101/280\r",
      "choosing prot 102/280\r",
      "choosing prot 103/280\r",
      "choosing prot 104/280\r",
      "choosing prot 105/280\r",
      "choosing prot 106/280\r",
      "choosing prot 107/280\r",
      "choosing prot 108/280\r",
      "choosing prot 109/280\r",
      "choosing prot 110/280\r",
      "choosing prot 111/280\r",
      "choosing prot 112/280\r",
      "choosing prot 113/280\r",
      "choosing prot 114/280\r",
      "choosing prot 115/280\r",
      "choosing prot 116/280\r",
      "choosing prot 117/280\r",
      "choosing prot 118/280\r",
      "choosing prot 119/280\r",
      "choosing prot 120/280\r",
      "choosing prot 121/280\r",
      "choosing prot 122/280\r",
      "choosing prot 123/280\r",
      "choosing prot 124/280\r",
      "choosing prot 125/280\r",
      "choosing prot 126/280\r",
      "choosing prot 127/280\r",
      "choosing prot 128/280\r",
      "choosing prot 129/280\r",
      "choosing prot 130/280\r",
      "choosing prot 131/280\r",
      "choosing prot 132/280\r",
      "choosing prot 133/280\r",
      "choosing prot 134/280\r",
      "choosing prot 135/280\r",
      "choosing prot 136/280\r",
      "choosing prot 137/280\r",
      "choosing prot 138/280\r",
      "choosing prot 139/280\r",
      "choosing prot 140/280\r",
      "choosing prot 141/280\r",
      "choosing prot 142/280\r",
      "choosing prot 143/280\r",
      "choosing prot 144/280\r",
      "choosing prot 145/280\r",
      "choosing prot 146/280\r",
      "choosing prot 147/280\r",
      "choosing prot 148/280\r",
      "choosing prot 149/280\r",
      "choosing prot 150/280\r",
      "choosing prot 151/280\r",
      "choosing prot 152/280\r",
      "choosing prot 153/280\r",
      "choosing prot 154/280\r",
      "choosing prot 155/280\r",
      "choosing prot 156/280\r",
      "choosing prot 157/280\r",
      "choosing prot 158/280\r",
      "choosing prot 159/280\r",
      "choosing prot 160/280\r",
      "choosing prot 161/280\r",
      "choosing prot 162/280\r",
      "choosing prot 163/280\r",
      "choosing prot 164/280\r",
      "choosing prot 165/280\r",
      "choosing prot 166/280\r",
      "choosing prot 167/280\r",
      "choosing prot 168/280\r",
      "choosing prot 169/280\r",
      "choosing prot 170/280\r",
      "choosing prot 171/280\r",
      "choosing prot 172/280\r",
      "choosing prot 173/280\r",
      "choosing prot 174/280\r",
      "choosing prot 175/280\r",
      "choosing prot 176/280\r",
      "choosing prot 177/280\r",
      "choosing prot 178/280\r",
      "choosing prot 179/280\r",
      "choosing prot 180/280\r",
      "choosing prot 181/280\r",
      "choosing prot 182/280\r",
      "choosing prot 183/280\r",
      "choosing prot 184/280\r",
      "choosing prot 185/280\r",
      "choosing prot 186/280\r",
      "choosing prot 187/280\r",
      "choosing prot 188/280\r",
      "choosing prot 189/280\r",
      "choosing prot 190/280\r",
      "choosing prot 191/280\r",
      "choosing prot 192/280\r",
      "choosing prot 193/280\r",
      "choosing prot 194/280\r",
      "choosing prot 195/280\r",
      "choosing prot 196/280\r",
      "choosing prot 197/280\r",
      "choosing prot 198/280\r",
      "choosing prot 199/280\r",
      "choosing prot 200/280\r",
      "choosing prot 201/280\r",
      "choosing prot 202/280\r",
      "choosing prot 203/280\r",
      "choosing prot 204/280\r",
      "choosing prot 205/280\r",
      "choosing prot 206/280\r",
      "choosing prot 207/280\r",
      "choosing prot 208/280\r",
      "choosing prot 209/280\r",
      "choosing prot 210/280\r",
      "choosing prot 211/280\r",
      "choosing prot 212/280\r",
      "choosing prot 213/280\r",
      "choosing prot 214/280\r",
      "choosing prot 215/280\r",
      "choosing prot 216/280\r",
      "choosing prot 217/280\r",
      "choosing prot 218/280\r",
      "choosing prot 219/280\r",
      "choosing prot 220/280\r",
      "choosing prot 221/280\r",
      "choosing prot 222/280\r",
      "choosing prot 223/280\r",
      "choosing prot 224/280\r",
      "choosing prot 225/280\r",
      "choosing prot 226/280\r",
      "choosing prot 227/280\r",
      "choosing prot 228/280\r",
      "choosing prot 229/280\r",
      "choosing prot 230/280\r",
      "choosing prot 231/280\r",
      "choosing prot 232/280\r",
      "choosing prot 233/280\r",
      "choosing prot 234/280\r",
      "choosing prot 235/280\r",
      "choosing prot 236/280\r",
      "choosing prot 237/280\r",
      "choosing prot 238/280\r",
      "choosing prot 239/280\r",
      "choosing prot 240/280\r",
      "choosing prot 241/280\r",
      "choosing prot 242/280\r",
      "choosing prot 243/280\r",
      "choosing prot 244/280\r",
      "choosing prot 245/280\r",
      "choosing prot 246/280\r",
      "choosing prot 247/280\r",
      "choosing prot 248/280\r",
      "choosing prot 249/280\r",
      "choosing prot 250/280\r",
      "choosing prot 251/280\r",
      "choosing prot 252/280\r",
      "choosing prot 253/280\r",
      "choosing prot 254/280\r",
      "choosing prot 255/280\r",
      "choosing prot 256/280\r",
      "choosing prot 257/280\r",
      "choosing prot 258/280\r",
      "choosing prot 259/280\r",
      "choosing prot 260/280\r",
      "choosing prot 261/280\r",
      "choosing prot 262/280\r",
      "choosing prot 263/280\r",
      "choosing prot 264/280\r",
      "choosing prot 265/280\r",
      "choosing prot 266/280\r",
      "choosing prot 267/280\r",
      "choosing prot 268/280\r",
      "choosing prot 269/280\r",
      "choosing prot 270/280\r",
      "choosing prot 271/280\r",
      "choosing prot 272/280\r",
      "choosing prot 273/280\r",
      "choosing prot 274/280\r",
      "choosing prot 275/280\r",
      "choosing prot 276/280\r",
      "choosing prot 277/280\r",
      "choosing prot 278/280\r",
      "choosing prot 279/280\r",
      "choosing prot 280/280"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<_io.TextIOWrapper name='/Users/zacharymcgrath/Desktop/nod2 data/not_truth_subset.fasta' mode='w' encoding='UTF-8'>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_keep = 280\n",
    "all_prots = {get_name(entry.description): entry for entry in fasta.read(whole_db)}\n",
    "all_prot_keys = list(all_prots.keys())\n",
    "\n",
    "saving = {}\n",
    "# get num_keep\n",
    "for i in range(num_keep):\n",
    "    print(f'\\rchoosing prot {i+1}/{num_keep}', end='')\n",
    "    \n",
    "    while not False:\n",
    "        \n",
    "        # pick a protein \n",
    "        selected = all_prot_keys[random.randint(0, len(all_prot_keys) - 1)]\n",
    "        \n",
    "        # see if its in saving or ts\n",
    "        if selected in ts or selected in saving:\n",
    "            continue\n",
    "            \n",
    "            \n",
    "        # flip it and add \"reverse\" to the description\n",
    "        entry = all_prots[selected]\n",
    "        \n",
    "        rev_seq = entry.sequence[::-1]\n",
    "        split_desc = entry.description.split('|')\n",
    "        split_desc[-1] = 'REVERSE_' + split_desc[-1]\n",
    "        new_desc = '|'.join(split_desc)\n",
    "        \n",
    "        saving[selected] = entry._replace(description=new_desc, sequence=rev_seq)\n",
    "        break\n",
    "            \n",
    "# write to file\n",
    "fasta.write([v for _, v in saving.items()], open(not_truth_set, 'w'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Run hyped search on both"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'spectra_folder': '/Users/zacharymcgrath/Desktop/nod2 data/filteredSpec/', 'database_file': '/Users/zacharymcgrath/Desktop/nod2 data/filteredNOD2.fasta', 'output_dir': '/Users/zacharymcgrath/Desktop/Experiment output/FP/truth/', 'min_peptide_len': 3, 'max_peptide_len': 30, 'tolerance': 20, 'precursor_tolerance': 3, 'peak_filter': 25, 'relative_abundance_filter': 0.0, 'verbose': True, 'DEBUG': False}\n",
      "Loading database...\n",
      "Done\n",
      "Loading spectra...\n",
      "Done\n",
      "On batch 1/1\n",
      "On protein 280/280 [100%]\n",
      "Sorting the set of protein masses...\n",
      "Creating an alignment for 29/1086 [2%]"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-7e0fd55b9c3c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtruth_run_params\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mrunner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtruth_run_params\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/runner.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(args)\u001b[0m\n\u001b[1;32m     51\u001b[0m         \u001b[0mrelative_abundance_filter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'relative_abundance_filter'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m         \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m         \u001b[0mDEBUG\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'DEBUG'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m     )\n\u001b[1;32m     55\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\nFinished search. Writting results to {}...'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'output_dir'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identification.py\u001b[0m in \u001b[0;36mid_spectra\u001b[0;34m(spectra_files, database_file, verbose, min_peptide_len, max_peptide_len, result_count, peak_filter, relative_abundance_filter, ppm_tolerance, precursor_tolerance, DEBUG)\u001b[0m\n\u001b[1;32m    416\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspectrum\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectra\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    417\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'\\rCreating an alignment for {i+1}/{len(spectra)} [{to_percent(i, len(spectra))}%]'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 418\u001b[0;31m         \u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mid_spectrum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mboundaries\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmz_mapping\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatched_masses_b\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmatched_masses_y\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mppm_tolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    419\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    420\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mresults\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/identification.py\u001b[0m in \u001b[0;36mid_spectrum\u001b[0;34m(spectrum, boundaries, db, mz_mapping, matched_masses_b, matched_masses_y, ppm_tolerance)\u001b[0m\n\u001b[1;32m    359\u001b[0m         \u001b[0mfiltered_y\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    360\u001b[0m         \u001b[0mppm_tolerance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mppm_tolerance\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 361\u001b[0;31m         \u001b[0mn\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    362\u001b[0m     )\n\u001b[1;32m    363\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/alignment/alignment.py\u001b[0m in \u001b[0;36mattempt_alignment\u001b[0;34m(spectrum, db, b_hits, y_hits, n, ppm_tolerance, precursor_tolerance, DEBUG, is_last)\u001b[0m\n\u001b[1;32m    241\u001b[0m         p_ms = [\n\u001b[1;32m    242\u001b[0m             \u001b[0mx\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 243\u001b[0;31m             \u001b[0malignment_utils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfill_in_precursor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgap\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallowed_gap\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mprecursor_tolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    244\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    245\u001b[0m         ]\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/alignment/alignment_utils.py\u001b[0m in \u001b[0;36mfill_in_precursor\u001b[0;34m(spectrum, sequence, db, gap, tolerance)\u001b[0m\n\u001b[1;32m    337\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mspectrum\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprecursor_mass\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mtheory_precrusor\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    338\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 339\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m__add_amino_acids\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspectrum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnum_off\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    340\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    341\u001b[0m     \u001b[0;31m# subtract amino acids:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/Layer_Research/hypedsearch/src/alignment/alignment_utils.py\u001b[0m in \u001b[0;36m__add_amino_acids\u001b[0;34m(spectrum, sequence, db, gap, tolerance)\u001b[0m\n\u001b[1;32m    112\u001b[0m                                 \u001b[0mnew_seq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0malign_overlaps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_left\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_right\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 114\u001b[0;31m                                 \u001b[0mnew_prec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgen_spectra\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_precursor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_seq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'('\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m')'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'-'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m                                 \u001b[0;31m# find the precursor distance, and if its close enough, keep it\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "print(truth_run_params)\n",
    "runner.run(truth_run_params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "runner.run(non_truth_run_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Compare and count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "truth_results_json = outputDir + 'truth/summary.json'\n",
    "non_truth_results_json = outputDir + 'not_truth/summary.json'\n",
    "\n",
    "truth_results = json.load(open(truth_results_json, 'r'))\n",
    "non_truth_results = json.load(open(non_truth_results_json, 'r'))\n",
    "\n",
    "# save them in a dictionary by their id so that we can compare\n",
    "ided_truth = {v['spectrum']['id'].replace('.pkl', ''): v for _, v in truth_results.items()}\n",
    "ided_non_truth = {v['spectrum']['id'].replace('.pkl', ''): v for _, v in non_truth_results.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "truth_greater = {}\n",
    "non_truth_greater = {}\n",
    "\n",
    "for k, v in ided_truth.items():\n",
    "    \n",
    "    if k not in ided_non_truth:\n",
    "        truth_greater[k] = v\n",
    "        \n",
    "    else:\n",
    "        nt_v = ided_non_truth[k]\n",
    "        \n",
    "        # see if we have anything in our alignments\n",
    "        if len(v['alignments']) == 0 and len(nt_v['alignments']) == 0:\n",
    "            continue\n",
    "            \n",
    "        elif len(v['alignments']) == 0 and len(nt_v['alignments']) > 0:\n",
    "            non_truth_greater[k] = v\n",
    "            continue\n",
    "            \n",
    "        elif len(v['alignments']) >0 and len(nt_v['alignments']) == 0:\n",
    "            truth_greater[k] = v \n",
    "            continue\n",
    "\n",
    "        # see which score was greater\n",
    "        best_truth = sorted(v['alignments'], key=lambda x: x['total_score'], reverse=True)[0]\n",
    "        best_non_truth = sorted(nt_v['alignments'], key=lambda x: x['total_score'], reverse=True)[0]\n",
    "        \n",
    "        if best_truth['total_score'] > best_non_truth['total_score']:\n",
    "            truth_greater[k] = (best_truth, best_non_truth)\n",
    "            \n",
    "        elif best_truth['sequence'] == best_non_truth['sequence']:\n",
    "            truth_greater[k] = (best_truth, best_non_truth)\n",
    "            \n",
    "        else:\n",
    "            non_truth_greater[k] = (best_truth, best_non_truth)\n",
    "            \n",
    "for k, v in ided_non_truth.items():\n",
    "    if k not in ided_truth and k not in truth_greater and k not in non_truth_greater:\n",
    "        non_truth_greater[k] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(truth_greater))\n",
    "print(len(non_truth_greater))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in non_truth_greater.items():\n",
    "    if isinstance(v, dict):\n",
    "        continue\n",
    "    print(f'Truth: {v[0][\"sequence\"]}, {v[0][\"total_score\"]} \\t non truth: {v[1][\"sequence\"]}, {v[1][\"total_score\"]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
